tmt/utils.py

```diff
@@ -374,20 +474,27 @@ def _run(
 
         # Handle the exit code, return output
         if process.returncode != 0:
-            if isinstance(command, (list, tuple)):
-                command = ' '.join(command)
             raise RunError(
                 message=f"Command returned '{process.returncode}'.",
                 command=command,
                 returncode=process.returncode,
                 stdout=stdout_thread.get_output(),
                 stderr=stderr_thread.get_output())
-        return stdout_thread.get_output() if join else (
+        return (stdout_thread.get_output(), None) if join else (
```

Comment:
Isn't there a better way to implement this join rather than always returning `None` in the second element? Couldn't the return type be a Union of `str` and `Tuple[str, str]` (+ `Optional` since `None` can be returned). But I guess this would cause problems in the calling code for mypy, right? So being consistent with the return types (by always returning a tuple) seems like the way to go.

Reply:
It is possible to change the type of return value to `Union[None, std, Tuple[str, str]]` (or any similarly structured type) - this is not a problem for mypy, but for humans. Once such a type is allowed, you need extra test for the type of return value before using it:

```python
distro = self.execute('cat /etc/os-release')[0].strip()
```

`self` here is a guest, and `execute()` method of `Guest` class ends with `return self.run(command, shell=False, **kwargs)`, therefore the type of its return value shall be same as `_run()` and `run()`. And mypy is going to report an error here because `None`, as one of the possible outcomes, is unscriptable.

```python
ret = self.execute('cat /etc/os-release')

if isinstance(ret, tuple):
  distro = ret[0].strip()
else:
  # what to do here?
```

These checks tend to spread throughout the code (a function that accepts "command output" should accept exactly the same type `execute()` or `run()` returns), and mypy **will** complain if you ever try to call a method/function that's incompatible with any member type of `Union`.

So, I took a shot, and reduced the number of possible outcomes to one, `Tuple[Optional[str], Optional[str]]`. It's open to discussion, of course. When it comes to a structure of return value, I for one prefer a single "kind", i.e. a `(stdout, None)` looks more appealing to me than getting `stdout` and `(stdout, stderr)` or just `None`, and I believe it leads to more readable code, easier to reason about - there's no space for doubt in my case, return value is **always** a tuple of two items, and these items are either strings (output) or `None`. I can't forget to test whether the returned value is `None` - it can never be, it's always a tuple -, it's easy to ignore either of outputs if I have no use for them, there will always be `[0]` or `[1]`.

So, mypy does not care that much, but if the return value can be of different types, additional tests will be needed before use, to both avoid runtime errors and to "convince" mypy that `[0]` in `[0].strip()` is indeed a string (from `("foo", "bar")` and not the first character of a stdout/stderr mix "foo\nbar\n" when `join=True` was used.

Reply:
How about returning namedtuple? One could access its first item by name so `run(..., join=True).stdout.strip()` will work.

Reply:
+1, a named tuple would be aligned with what I described above, it'd stick to the "one return type only" approach, just adding names for better readability.

Reply:
Yeah, "one return type only" is the best.

Reply:
Changed to a named tuple in 518fecb.